Lib/dataclasses.py

```diff
@@ -1262,18 +1281,33 @@ def _add_slots(cls, is_frozen, weakref_slot):
 
     # And finally create the class.
     qualname = getattr(cls, '__qualname__', None)
-    cls = type(cls)(cls.__name__, cls.__bases__, cls_dict)
+    newcls = type(cls)(cls.__name__, cls.__bases__, cls_dict)
     if qualname is not None:
-        cls.__qualname__ = qualname
+        newcls.__qualname__ = qualname
 
     if is_frozen:
         # Need this for pickling frozen classes with slots.
         if '__getstate__' not in cls_dict:
-            cls.__getstate__ = _dataclass_getstate
+            newcls.__getstate__ = _dataclass_getstate
         if '__setstate__' not in cls_dict:
-            cls.__setstate__ = _dataclass_setstate
-
-    return cls
+            newcls.__setstate__ = _dataclass_setstate
+
+    # Fix up any closures which reference __class__.  This is used to
+    # fix zero argument super so that it points to the correct class
+    # (the newly created one, which we're returning) and not the
+    # original class.
+    for member in newcls.__dict__.values():
+        # If this is a wrapped function, unwrap it.
+        member = inspect.unwrap(member)
+
+        if isinstance(member, types.FunctionType):
+            _update_func_cell_for__class__(member, cls, newcls)
+        elif isinstance(member, property):
+            _update_func_cell_for__class__(member.fget, cls, newcls)
+            _update_func_cell_for__class__(member.fset, cls, newcls)
+            _update_func_cell_for__class__(member.fdel, cls, newcls)
```

Comment:
I think technically we could break out of the loop as soon as we've successfully updated one cell, because all the closures for a given class will always share a single `__class__` cell. This may improve performance noticeably for larger classes?

Reply:
Good point. Will change.